jupyterhub/oauth/provider.py

```diff
@@ -547,35 +549,51 @@ def validate_scopes(self, client_id, scopes, client, request, *args, **kwargs):
             - Resource Owner Password Credentials Grant
             - Client Credentials Grant
         """
+
         orm_client = (
             self.db.query(orm.OAuthClient).filter_by(identifier=client_id).one_or_none()
         )
         if orm_client is None:
             app_log.warning("No such oauth client %s", client_id)
             return False
-        client_allowed_roles = {role.name: role for role in orm_client.allowed_roles}
+
         # explicitly allow 'identify', which was the only allowed scope previously
         # requesting 'identify' gets no actual permissions other than self-identification
-        client_allowed_roles.setdefault('identify', None)
-        roles = []
-        requested_roles = set(scopes)
-        disallowed_roles = requested_roles.difference(client_allowed_roles)
-        if disallowed_roles:
+        scopes = set(scopes)
+        scopes.discard("identify")
+
+        # TODO: handle roles->scopes transition
+        # at this point, 'scopes' _may_ be roles
```

Comment:
Can you make this comment more explicit? To clarify my confusion, "at this point" could refer to the upcoming Jupyterhub version or the current WIP state of this PR etc.

Reply:
For compatibility, I want to make a release that doesn't break OAuth as it is. Right now, our OAuth API says: "When requesting scopes, use role names as the requested scopes, not actual scopes." This is covered in the test `test_services_auth.py:test_oauth_service_roles`. After this change, it _expects_ scopes, but _allows_ roles. I think this role->scope conversion should be _deprecated_, though I'm not 100% sure.

Assigning permissions via roles as permission-bundles is still logical and convenient (e.g. the tokens REST API still accepts roles and I'm not sure it should be deprecated), so I don't have a firm idea where exactly in user-facing APIs it should be deprecated, vs just changing the underlying storage mechanism by resolving roles->scopes at assignment time.

Reply:
(notably on my todo:  test_services_auth.py:test_oauth_service_roles doesn't test requesting scopes instead of roles, so the deprecated behavior is tested, but the new behavior is not, yet)